tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/SampleMetadataProvider.java

```diff
@@ -0,0 +1,21 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics.registry;
+
+import java.util.function.Supplier;
+
+public interface SampleMetadataProvider extends Supplier<SampleMetadata> {}
```

Comment:
Should we name this something about exemplars given that's the intended use case?

API design nit: Instead of extending Supplier, I prefer to declare a method, in this case that might look something like:

```java
public interface SampleMetadataProvider {
    SampleMetadata collect();
}
```

In practice I think we may want this to return an Optional so we're not forced to return something if, for example, there's no traceId.

I think we'll also want to make the SampleMetadata type generic so that tritium isn't necessarily responsible for defining what's collected

Reply:
Yeah that's fair. I had been trying to avoid the term "Exemplars" as product-specific and instead just using Sample, but it seems like it's defined by OpenTelemetry too, which I hadn't seen (https://opentelemetry.io/docs/specs/otel/metrics/data-model/#exemplars) so having Exemplars in the name makes sense to me.

Reply:
Something like https://github.com/palantir/tritium/pull/1992/commits/5cf5d7febeacc359ce10e0d72f3e9841993875f8 for the generic type?

Reply:
I think we need to get a bit clever with how we handle the generic type. When we report metrics via `registry.forEachMetric(name, metric -> {})` we won't have the generic type info anymore.

I could see something like leveraging the `ExemplarMetadataProvider<T>` for extraction, that way we end up calling something along these lines:

```
public interface ExemplarsCapture {
    <T> List<T> getExemplars(ExemplarMetadataProvider<T> provider);
}
```

Internally we would be able to check whether or not the exemplar values are sourced from this provider otherwise return an empty list.
This way the reservoir itself needn't be generic, and if someone were to register a different histogram from another metadata provider, it would be ignored without causing issues in the reporting code.

Reply:
But the consumer (eg sls-logging) won't have access to the provider to be able to pass it in. Do you mean to expose it and pass it in to check that the parameterized type (T) maps the exemplar collection we have internally? Or to have the WeightedSnapshot actually have some metadata on which provider was used?